### ISSUE TITLE
False Positive: `needless_raw_string_hashes` when combined with `const_format::formatcp`

### ISSUE THREAD

**Issue by hoijui** [rust-lang/rust-clippy#11748]
### Summary

When using `formatcp!` from `const_format`,
Some FPs happen, following (to me) strange rules.

### Lint Name

needless_raw_string_hashes

### Reproducer

Using `const_format = "0.2"` in Cargo.toml.

I tried this code:

```rust
#![deny(clippy::needless_raw_string_hashes)]

use const_format::formatcp;

const C: &str = "c";

fn main() {
    println!(
        "{}",
        formatcp!(
            // NOTE warning -> false positive!
            r#"a "b" {C} d"#
        )
    );

    println!(
        "{}",
        formatcp!(
            // no warning -> correct
            r#"a "b" {C}"#
        )
    );
}
```

I saw this happen:

```
error: unnecessary hashes around raw string literal
  --> clippy-raw-string-lit-test/src/main.rs:15:13
   |
15 |             r#"a "b" {C} d"#
   |             ^^^^^^^^^^^^^^^^
```

I expected to see this happen:

### Version

```text
rustc 1.75.0-nightly (9d83ac217 2023-10-31)
binary: rustc
commit-hash: 9d83ac217957eece2189eccf4a7232caec7232ee
commit-date: 2023-10-31
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.3
```

**Comment by y21** [comment]
It looks like what happens here is that the `formatcp!` macro "splits" up parts of the string literal passed to it into smaller string literals: `r#"a "b" "#` and `r" d"#`. The latter does not need hashes.

In the expanded code as clippy sees it, both string literals have the same span (the span of the whole string literal), so any warning/suggestion we make to either string applies to the string literal that the user typed.

Might be one of those cases where the macro author should emit a `#[allow(clippy::needless_raw_string_hashes)]` as part of its expansion?

**Comment by hoijui** [comment]
... sounds good!
what you suggest @y21, is already in [this PR](https://github.com/rodrimati1992/const_format_crates/pull/54) for an other lint for the same macro.

**Comment by hoijui** [comment]
Reported in [an issue at the `const_format` crate](https://github.com/rodrimati1992/const_format_crates/issues/56).
Should we close this here or leave open till it is fixed there?
